cirq/circuits/text_diagram_drawer_test.py

```diff
@@ -159,3 +191,165 @@ def test_multiline_entries():
+    assert orig_drawer == copy_drawer
+
+    copy_drawer.write(0, 1, 'new_entry')
+    assert copy_drawer != orig_drawer
+
+    copy_drawer = orig_drawer.copy()
+    copy_drawer.vertical_line(2, 1, 3)
+    assert copy_drawer != orig_drawer
+
+    copy_drawer = orig_drawer.copy()
+    copy_drawer.horizontal_line(2, 1, 3)
+    assert copy_drawer != orig_drawer
+
+    copy_drawer = orig_drawer.copy()
+    copy_drawer.force_horizontal_padding_after(1, 4)
+    assert copy_drawer != orig_drawer
+
+    copy_drawer = orig_drawer.copy()
+    copy_drawer.force_vertical_padding_after(1, 4)
+    assert copy_drawer != orig_drawer
+
+
+def test_drawer_stack():
+    d = TextDiagramDrawer()
+    d.write(0, 0, 'A')
+    d.write(1, 0, 'B')
+    d.write(1, 1, 'C')
+    dd = TextDiagramDrawer()
+    dd.write(0, 0, 'D')
+    dd.write(0, 1, 'E')
+    dd.write(1, 1, 'F')
+
+    vstacked = TextDiagramDrawer.vstack((dd, d))
+    expected = """
+D
+
+E F
+
+A B
+
+  C
+    """.strip()
+    assert_has_rendering(vstacked, expected)
+
+    hstacked = TextDiagramDrawer.hstack((d, dd))
+    expected = """
+A B D
+
+  C E F
+    """.strip()
+    assert_has_rendering(hstacked, expected)
+
+    d.force_horizontal_padding_after(0, 0)
+
+    with pytest.raises(ValueError):
+        TextDiagramDrawer.vstack((dd, d))
+
+    dd.force_horizontal_padding_after(0, 0)
+    expected = """
+D
+
+EF
+
+AB
+
+ C
+    """.strip()
+    vstacked = TextDiagramDrawer.vstack((dd, d))
+    assert_has_rendering(vstacked, expected)
+
+    d.force_vertical_padding_after(0, 0)
+    with pytest.raises(ValueError):
+        print(d.vertical_padding)
+        print(dd.vertical_padding)
+        TextDiagramDrawer.hstack((d, dd))
+
+
+    dd.force_vertical_padding_after(0, 0)
+    expected = """
+AB D
+ C EF
+    """.strip()
+    hstacked = TextDiagramDrawer.hstack((d, dd))
+    assert_has_rendering(hstacked, expected)
+
+    d.force_horizontal_padding_after(0, 0)
+    dd.force_horizontal_padding_after(0, 2)
+    d.force_vertical_padding_after(0, 1)
+    dd.force_vertical_padding_after(0, 3)
+
+    with pytest.raises(ValueError):
+        TextDiagramDrawer.vstack((d, dd))
+
+
+    vstacked = TextDiagramDrawer.vstack((dd, d), padding_resolver=max)
+    expected = """
+D
+
+
+
+E  F
+
+A  B
+
+   C
+    """.strip()
+    assert_has_rendering(vstacked, expected)
+
+    hstacked = TextDiagramDrawer.hstack((d, dd), padding_resolver=max)
+    expected = """
+AB D
+
+
+
+ C E  F
+    """.strip()
+    assert_has_rendering(hstacked, expected)
+
+    vstacked_min = TextDiagramDrawer.vstack((dd, d), padding_resolver=min)
+    expected = """
+D
+
+
+
+EF
+
+AB
+
+ C
+    """.strip()
+    assert_has_rendering(vstacked_min, expected)
+
+    hstacked_min = TextDiagramDrawer.hstack((d, dd), padding_resolver=min)
+    expected = """
+AB D
+
+ C E  F
+    """.strip()
+    assert_has_rendering(hstacked_min, expected)
+
+def test_drawer_eq():
+    assert TextDiagramDrawer().__eq__(23) == NotImplemented
```

Comment:
Use `EqualsTester` to properly exercise the eq, neq, etc methods.

Reply:
Done